File: backend/utils/riderBalances.js
const { Op } = require('sequelize');
const { Rider, Transaction } = require('../models');

const OUTSTANDING_TRANSACTION_STATUSES = ['pending', 'approved'];

async function getOutstandingBalanceMap(riderIds = []) {
  if (!riderIds.length) {
    return new Map();
  }

  const totals = await Transaction.findAll({
    attributes: [
      [Transaction.sequelize.col('rider_id'), 'riderId'],
      [Transaction.sequelize.fn('SUM', Transaction.sequelize.col('amount')), 'balance'],
    ],
    where: {
      riderId: {
        [Op.in]: riderIds,
      },
      status: {
        [Op.in]: OUTSTANDING_TRANSACTION_STATUSES,
      },
    },
    group: [Transaction.sequelize.col('rider_id')],
    raw: true,
  });

  return new Map(
    totals.map((entry) => [
      Number(entry.riderId ?? entry.rider_id),
      Number(entry.balance || 0),
    ])
  );
}

async function syncRiderBalance(riderId) {
  const balanceMap = await getOutstandingBalanceMap([riderId]);
  const balance = balanceMap.get(Number(riderId)) || 0;

  await Rider.update(
    { currentBalance: balance },
    {
      where: { id: riderId },
    }
  );

  return balance;
}

async function syncAllRiderBalances() {
  const riders = await Rider.findAll({ attributes: ['id'], raw: true });
  const balanceMap = await getOutstandingBalanceMap(riders.map((rider) => rider.id));

  await Promise.all(
    riders.map((rider) =>
      Rider.update(
        { currentBalance: balanceMap.get(Number(rider.id)) || 0 },
        {
          where: { id: rider.id },
        }
      )
    )
  );
}

module.exports = {
  getOutstandingBalanceMap,
  syncRiderBalance,
  syncAllRiderBalances,
  OUTSTANDING_TRANSACTION_STATUSES,
};

File: backend/app/utils/rider_balances.py
from sqlalchemy import func

from app.database.db import db
from models import Rider, Station, Transaction


OUTSTANDING_TRANSACTION_STATUSES = {"pending", "approved"}


def get_outstanding_balance_map(
    rider_ids=None,
    company_id=None,
    company_name=None,
    station_id=None,
):
    rider_ids = rider_ids or []

    if not rider_ids:
        return {}

    query = (
        db.session.query(
            Transaction.rider_id.label("rider_id"),
            func.coalesce(func.sum(Transaction.amount), 0).label("balance"),
        )
        .filter(Transaction.rider_id.in_(rider_ids))
        .filter(Transaction.status.in_(OUTSTANDING_TRANSACTION_STATUSES))
    )

    if company_id or company_name:
        query = query.join(Station, Station.id == Transaction.station_id)
        if company_id:
            query = query.filter(Station.company_id == company_id)
        elif company_name:
            query = query.filter(Station.company_name == company_name)

    if station_id:
        query = query.filter(Transaction.station_id == station_id)

    rows = query.group_by(Transaction.rider_id).all()

    return {row.rider_id: float(row.balance or 0) for row in rows}


def sync_rider_balance(rider_id):
    balance = get_outstanding_balance_map([rider_id]).get(rider_id, 0)
    rider = Rider.query.get(rider_id)

    if rider:
        rider.current_balance = balance
        db.session.commit()

    return balance
